src/main/java/com/aws/greengrass/clientdevices/auth/iot/Thing.java

```diff
@@ -85,37 +85,43 @@ public Thing clone() {
      * @param certificateId Certificate ID to attach
      */
     public void attachCertificate(String certificateId) {
-        if (attachedCertificateIds.addIfAbsent(certificateId)) {
-            modified = true;
+        if (attachedCertificateIds.containsKey(certificateId)) {
+            return;
         }
+        attachedCertificateIds.put(certificateId, Instant.now());
+        modified = true;
     }
 
     /**
      * Detach a certificate ID.
      * @param certificateId Certificate ID to detach
      */
     public void detachCertificate(String certificateId) {
-        if (attachedCertificateIds.remove(certificateId)) {
+        if (attachedCertificateIds.remove(certificateId) != null) {
             modified = true;
         }
     }
 
     /**
      * Returns copy of attached certificate IDs.
-     * </p>
-     * This list cannot be modified directly. Refer to {@link #attachCertificate(String) attachCertificate} and
+     * <p>
+     * This map should not be modified directly. Refer to {@link #attachCertificate(String) attachCertificate} and
      * {@link #detachCertificate(String) detachCertificate}
-     *
+     * </p>
      * @return Certificate IDs
      */
-    public List<String> getAttachedCertificateIds() {
-        return new ArrayList<>(attachedCertificateIds);
+    public Map<String, Instant> getAttachedCertificateIds() {
```

Comment:
I don't know if this can be improved, but this breaks encapsulation and exposes how we internally structure this class. Ideally we are not returning the attached certificates but rather asking this class to do something for us which is what the [tell don't ask principle](https://martinfowler.com/bliki/TellDontAsk.html#:~:text=Tell%2DDon't%2DAsk,an%20object%20what%20to%20do.) is all about. 

It is important because it allows us to change how we store data inside of this class without exposing the externals giving us the freedom to make any change we want inside of this class without affecting the external world.

Reply:
Agreed. How about the `Thing` internally tracks the Instant for each certId ( with `Map<String, Instant>`) and expose only the list of certIds to the external world?

Reply:
lol, or the other way around: Thing only tracks list of attached certIds and something else (`ThingRegistry`) tracks certId <-> Instant mapping in `ThingV1` dto

Reply:
I think we're missing a layer here. These are low level primitives within the iot domain, but external callers don't care about these details.

We want callers to be able to say "authenticate this connection" and let us figure out what that means in terms of Things and Certificates being attached.

Reply:
Regardless of the layer I think we should not expose the internals. Callers can still do what they want and interact with the thing through its API.

Reply:
Can refactor this later. Right now we have callers depending on this API. Removing it breaks a bunch of stuff